src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs

```diff
@@ -88,33 +92,47 @@ public new void RemoveMetadataReference(string referencePath)
             base.RemoveMetadataReference(referencePath);
         }
 
-        public void AddProjectReference(IProjectContext project, MetadataReferenceProperties properties)
+        public void AddProjectReference(IWorkspaceProjectContext project, MetadataReferenceProperties properties)
         {
+            var abstractProject = GetAbstractProject(project);
+
             // AbstractProject and ProjectTracker track project references using the project bin output path.
             // Setting the command line arguments should have already set the output file name and folder.
             // We fetch this output path to add the reference.
-            var referencedProject = this.ProjectTracker.GetProject(project.Id);
+            var referencedProject = this.ProjectTracker.GetProject(abstractProject.Id);
             var binPathOpt = referencedProject.TryGetBinOutputPath();
             if (!string.IsNullOrEmpty(binPathOpt))
             {
                 AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(binPathOpt, properties);
             }
         }
 
-        public void RemoveProjectReference(IProjectContext project)
+        public void RemoveProjectReference(IWorkspaceProjectContext project)
+        {
+            var referencedProject = GetAbstractProject(project);
+
+            // AbstractProject and ProjectTracker track project references using the project bin output path.
+            // We fetch this output path to remove the reference.
```

Comment:
Base abstract project does bunch of extra stuff in [RemoveMetadataReference](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/AbstractProject.cs,486), such as updates `_metadataFileNameToConvertedProjectReference`, which is needed to be done here. I found this out as the test was throwing during AbstractProject.Disconnect because `_metadataFileNameToConvertedProjectReference` had stale entries with previous code. We need to clean up the AbstractProject code when dealing with metadata/project references and converting between them.